README
KRISP version 2.1.0
===================


The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.


DESCRIPTION

This is perl module that use libkrisp.

KRISP provides network calcluration and searching ISP/country information
for given ip address.


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install


DEPENDENCIES

This module requires these libraries:

  Over libkrisp 3.1.0.

  If you use libkrisp under 3.0 (for example 2.x.x), use
  v1 branch. Head branch is required over libkrisp 3.1.0.

  If you use libkrisp 3.0.x, use KRISP module 2.0.x version.


USAGE

See also ```man 3 KRISP``` or wiki page (https://github.com/Joungkyun/perl-KRISP/wiki)


COPYRIGHT AND LICENCE

Copyright (C) 2016 by JoungKyun.Kim <http://oops.org>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU LESSER GENERAL PUBLIC LICENSE as published
by the Free Software Foundation; either version 2.1 of the License.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU LESSER GENERAL PUBLIC LICENSE for more details.

You should have received a copy of the GNU LESSER GENERAL PUBLIC LICENSE
along with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.



Changes
Revision history for Perl extension KRISP.


2.1.1
	* Tue Jan 24 2012 JoungKyun.Kim <http://oops.org>
	- fixed underflow on ip2long on 64bit machine
	- fixed segfault on 64bit gcc4

	* Wed Apr 27 2011 JoungKyun.Kim <http://oops.org>
	- 2.1.1 tree start


2.1.0
	* Sat Sep 11 2010 JoungKyun.Kim <http://oops.org>
	- linked libkrisp version at least 3.1.0
	- add set_debug method
	- add set_mtime_interval method


2.0.1
	* Mon Aug  9 2010 JoungKyun.Kim <http://oops.org>
	- changed license from GPLv2 to LGPL which current and all
	  old versions.
	- linked libkrisp version at least 3.0.2

	* Tue Aug  3 2010 JoungKyun.Kim <http://oops.org>
	- changed path of libkrisp header files


2.0.0
	* Sun Jul  4 2010 JoungKyun.Kim <http://oops.org>
	- don't use swig wrapper any more
	- perl native c module

	* Tue Jun 29 2010 JoungKyun.Kim <http://oops.org>
	- rewrite all code
	  h2xs -A -n KRISP (with 1.23)

	* Mon Jun 21 2010 JoungKyun.Kim <http://oops.org>
    - required over libkrisp 3.0.0


1.2.0
	* Fri Dec  1 2006 JoungKyun.Kim <http://oops.org>
    - release 1.2.0

	* Wed Nov 29 2006 JoungKyun.Kim <http://oops.org>
    - support over libkrisp 2.1.0
    - support userdb support


1.1.1
	* Thu Sep 15 2006 JoungKyun.Kim <http://oops.org>
	- add geoip city flag in new declear
	- 1.1.1 release


1.1.0
	* Thu Sep  9 2006 JoungKyun.Kim <http://oops.org>
	- required over libkrisp 2.0.0
	- support several geoip extensions
	- release 1.1.0


1.0.0
	* Mon Jun 26 2006 JoungKyun.Kim <http://oops.org>
	- 1.0.0 release


krisp.c
/*
 * $Id: krisp.c,v 1.2 2006-09-07 14:49:03 oops Exp $
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <krisp.h>

extern char dberr[1024];
char krerr[1024];

char * krisp_version_pl (void) {
	return krisp_version ();
}

char * krisp_uversion_pl (void) {
	return krisp_uversion ();
}

char * krisp_error_pl (void) {
	return krerr;
}

KR_API * krisp_open_pl (char *datafile) {
	KR_API *db;

	struct stat f;
	int r;

	r = stat (datafile, &f);

	if ( r == -1 ) {
		sprintf (krerr, "datafile not found : %s", datafile);
		return NULL;
	}

	if ( f.st_size < 1 ) {
		sprintf (krerr, "datafile size is zero: %s", datafile);
		return NULL;
	}

	db = (KR_API *) malloc (sizeof (KR_API));

	if ( kr_open (db, datafile) ) {
		strcpy (krerr, dberr);
		free (db);
		return NULL;
	}

	return db;
}

char * krisp_search_pl (KR_API *db, char *host) {
	KRNET_API isp;
	static char ret[1024];

	if ( strlen (host) > 255 ) {
		strncpy (isp.ip , host, 255);
		isp.ip[255] = 0;
	} else {
		strcpy (isp.ip, host);
	}

	if ( kr_search (&isp, db) ) {
		return NULL;
	}

	sprintf (ret, "%s:%s:%s:%s:%s:%s:%s:%s:%s:%s",
			isp.key, isp.ip, isp.netmask, isp.network, isp.broadcast,
			isp.icode, isp.iname, isp.gcode, isp.gname, isp.gcity);

	return ret;
}

void krisp_close_pl (KR_API *db) {
	if ( db != NULL ) {
		kr_close (db);
		free (db);
	}
}

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */


version.h
/*
 * Copyright (C) 2016 by JoungKyun.Kim <http://oops.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU LESSER GENERAL PUBLIC LICENSE as published
 * by the Free Software Foundation; either version 2.1 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU LESSER GENERAL PUBLIC LICENSE for more details.
 *
 * You should have received a copy of the GNU LESSER GENERAL PUBLIC LICENSE
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#define MODVER "2.1.1"
#define MODUVER "002001001"

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */


Makefile.PL
use 5.008008;
use ExtUtils::MakeMaker;

my @SEARCHPATHS = (
	"/bin", "/sbin",
	"/usr/bin", "/usr/sbin",
	"/usr/local/bin", "/usr/local/sbin",
	"/usr/local/krisp/bin", "/usr/local/libkrisp/bin"
);

my $krispconfig;

foreach my $i (@SEARCHPATHS) {
	if ( -f "$i/krisp-config" ) {
		$krispconfig = "$i/krisp-config";
		last;
	}
}

if ( $krispconfig eq "" ) {
	printf "ERROR: krisp-config is not found\n";
	exit (1);
}

my $library = `$krispconfig --libs`;
#my $library = '-lkrisp';
my $cputype = `uname -m`;
my $ostype  = `uname -s`;
my $define = '';
my $inc = '';
my $optimize = 2;
my $march = '';

$library = `$krispconfig --link-ld 2> /dev/null`;
if ( ! $library ) {
	$library = `$krispconfig --libs 2> /dev/null`;
	$library = '-lkrisp' if ( ! $library );
}

$inc = `$krispconfig --includes 2> /dev/null`;
$define = `$krispconfig --defs 2> /dev/null`;


chomp ($library);
chomp ($inc);
chomp ($cputype);
chomp ($ostype);
$march = "-march=$cputype" if ( $cputype ne "x86_64" );

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'KRISP',
	DISTNAME          => 'KRISP',
	OPTIMIZE          => "-O$optimize $march",
    VERSION_FROM      => 'lib/KRISP.pm', # finds $VERSION
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/KRISP.pm', # retrieve abstract from module
       AUTHOR         => 'JoungKyun.Kim <https://joungkyun.github.com>') : ()),
	TYPEMAPS          => ['map/typemap'],
    LIBS              => ["$library"], # e.g., '-lm'
    DEFINE            => "$define", # e.g., '-DHAVE_SOMETHING'
    INC               => "-I. $inc", # e.g., '-I. -I/usr/include/other'
	dist              => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
	# Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
);

# Local variables:
# tab-width: 4
# c-basic-offset: 4
# End:
# vim600: noet sw=4 ts=4 fdm=marker
# vim<600: noet sw=4 ts=4


lib/KRISP.pm
package KRISP;

# Copyright (C) 2016 by JoungKyun.Kim <http://oops.org>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU LESSER GENERAL PUBLIC LICENSE as published
# by the Free Software Foundation; either version 2.1 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU LESSER GENERAL PUBLIC LICENSE for more details.
#
# You should have received a copy of the GNU LESSER GENERAL PUBLIC LICENSE
# along with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

# Local variables:
# tab-width: 4
# c-basic-offset: 4
# End:
# vim600: noet sw=4 ts=4 fdm=marker
# vim<600: noet sw=4 ts=4

use 5.008008;
use strict;
use warnings;

use constant false => 0;
use constant true  => 1;

require Exporter;

our @ISA = qw(Exporter);

our %EXPORT_TAGS = ( 'all' => [ qw( ) ] );
our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
our @EXPORT = qw( );

our $VERSION = '2.1.0';
my $revision = '$Id$';

require XSLoader;
XSLoader::load('KRISP', $VERSION);

# Preloaded methods go here.

# ------- Objects --------

=head1 NAME

KRISP - Perl extension for libkrisp

=head1 SYNOPSIS

    use KRISP;
    my $kr = new KRISP;

    # network convert
    my $start = '192.168.1.3';
    my $end   = '192.168.1.100';

    my $longip    = $kr->ip2long ($start);
    my $ipv4      = $kr->long2ip ($longip);
    my $netmask   = $kr->netmask ($start, $end);
    my $network   = $kr->network ($start, $mask);
    my $broadcast = $kr->broadcast ($start, $mask);
    my $prefix    = $kr->mask2prefix ($mask);
    my $mask      = $kr->prefix2mask ($prefix);

    # search isp/nation information
    my $err;
    my $db = $kr->open ('', $err);
    $kr->set_debug ($db, KRISP::true);
    my $isp = $kr->search ($db, 'some.domain.name', $err);
    my $isp_ex = $kr->search_ex ($db, 'some.domain.name', 'krisp', $err);
    $kr->close ($db);

=head1 DESCRIPTION

KRISP provides network cacluration and searching isp/nation
information for given ip address.

=head1 Constructor

=head2 new KRISP

Returns a new KRISP object.

=cut

sub new {
	my $self = {};
	return bless $self;
}


1;
__END__

=head1 Methods

=head2 $kr->mod_version ()

Returns version of this modules.

=head2 $kr->mod_uversion ()

Returns numeric version of this modules.

=head2 $kr->version ()

Returns version of libkrisp that linked.

=head2 $kr->mod_uversion ()

Returns numeric version of libkrisp that linked.

=head2 $kr->ip2long ($ipv4_address)

Calculates unsigned long value about given IPv4 address.

=over 4

=item ipv4_address

IPv4 address

=back

=head2 $kr->long2ip ($long_ip)

Calculates IPv4 address about given unsigend long address.

=over 4

=item long_ip

Unsigned long ip address

=back

=head2 $kr->netmask ($start, $end);

Calculates network mask that give range between $start and $end.

=over 4

=item start

First IPv4 address of Range

=item end

Last IPv4 address of Range

=back

=head2 $kr->mask2prefix ($mask);

Calculates network prefix about given network mask.

=over 4

=item mask

Network mask address

=back

=head2 $kr->prefix2mask ($prefix);

Calculates network mask about given network prefix.

=over 4

=item prefix

Network prefix

=back

=head2 $kr->network ($ipv4, $mask);

Calculates network address about given IP address and network mask.

=over 4

=item $ipv4

IPv4 address

=item $mask

Network mask

=back

=head2 $kr->broadcast ($ipv4, $mask);

Calculates broadcast address about given IP address and network mask.

=over 4

=item $ipv4

IPv4 address

=item $mask

Network mask

=back

=head2 $kr->open ([$database[, $error]]);

Open the krisp database handle.

=over 4

=item database (optional)

Specifies krisp database file path. If don't given this value,
libkrisp is executes by default value that set libkrisp compile
times.

=item error (optional)

Set this variable, if occures error on open method, allocates
error messages on this variable.

=back

=head2 $kr->search ($open_handle, $host[, $error]);

Returns network information about given host that includes
network information, nation information and isp information.

=over 4

=item open_handle

Return value of open method.

=item host

Searching host

=item error (optional)

Set this variable, if occures error on open method, allocates
error messages on this variable.

=back

=head2 $kr->search_ex ($open_handle, $host, $table[, $error]);

If you want to search your table that makes by you in krisp database,
you can use this method.

Search on given table and returns network information about given
host that includes network information, nation information and
isp information.

=over 4

=item open_handle

Return value of open method.

=item host

Searching host

=item table

Searching table on krisp database

=item error (optional)

Set this variable, if occures error on open method, allocates
error messages on this variable.

=back

=head2 $kr->close ($open_handle);

Closes krisp database handle.

=over 4

=item open_handle

Return value of open method.

=back

=head2 $kr->set_debug ($open_handle, $boolean)

Print libkrisp debug messages to standard out.

=over 4

=item open_handle

Return value of open method.

=item boolean (optional)

Default value is KRISP::true (1). Set KRISP::true or 1, search
method is printed debug messages to standard out. If you want to
no debug messages, set KRISP::false or 0.

If this method has not been called, Default value will be meaningless.

=back

=head2 $kr->set_mtime_interval ($open_handle, $seconds)

From libkrisp 3.1.0, libkrisp is periodically to detect database
changes. If libkrisp detects database changes, internally to close
database and will open database again.

libkrip checks change of database per 3600 seconds.
This method is enabled to change check-peroid of database.

=over 4

=item open_handle

Return value of open method.

=item seconds

Peroid of database check. If you set 0, libkrisp don't check changes
of database.

=back

=head1 AUTHOR

JoungKyun.Kim , E<lt>admin@oops.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2016 by JoungKyun.Kim

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU LESSER GENERAL PUBLIC LICENSE as published
by the Free Software Foundation; either version 2.1 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU LESSER GENERAL PUBLIC LICENSE for more details.

You should have received a copy of the GNU LESSER GENERAL PUBLIC LICENSE
along with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


=cut
